mlflow/transformers.py

```diff
@@ -0,0 +1,909 @@
+                                usage with pyfunc inference.
+
+    :param task: The transformers-specific task type of the model. These strings are utilized so
+                 that a pipeline can be created with the appropriate internal call architecture
+                 to meet the needs of a given model. If this argument is not specified, the
+                 pipeline utilities within the transformers library will be used to infer the
+                 correct task type. If the value specified is not a supported type within the
+                 version of transformers that is currently installed, an Exception will be thrown.
+    :param model_card: An Optional `ModelCard` instance from `huggingface-hub`. If provided, the
+                       contents of the model card will be saved along with the provided
+                       `transformers_model`. If not provided, an attempt will be made to fetch
+                       the card from the base pretrained model that is provided (or the one that is
+                       included within a provided `Pipeline`).
+
+                       .. Note:: In order for a ModelCard to be fetched (if not provided),
+                                 the huggingface_hub package must be installed and the version
+                                 must be >=0.10.0
+
+    :param code_paths: A list of local filesystem paths to Python file dependencies (or directories
+                       containing file dependencies). These files are *prepended* to the system
+                       path when the model is loaded.
+    :param mlflow_model: An MLflow model object that specifies the flavor that this model is being
+                         added to.
+    :param signature: A Model Signature object that describes the input and output Schema of the
+                      model. The model signature can be inferred using `infer_signature` function
+                      of `mlflow.models.signature`.
+    :param input_example: An example of valid input that the model can accept. The example can be
+                          used as a hint of what data to feed the model. The given example will be
+                          converted to a `Pandas DataFrame` and then serialized to JSON using the
+                          `Pandas` split-oriented format. Bytes are base64-encoded.
+    :param pip_requirements: {{ pip_requirements }}
+    :param extra_pip_requirements: {{ extra_pip_requirements }}
+    :param conda_env: {{ conda_env }}
+    :param metadata: Custom metadata dictionary passed to the model and stored in the MLmodel file.
+
+                     .. Note:: Experimental: This parameter may change or be removed in a future
+                                             release without warning.
+
+    :param kwargs: Optional additional configurations for transformers serialization.
+    :return: None
+    """
+    import transformers
+
+    _validate_transformers_model_dict(transformers_model)
```

Comment:
Can we define a named tuple for a dict `transformers_model`?


```python
class _TransformerModel(NamedTuple):
    # Internal representation of a dict transformers_model
    model
    ...

transformers_model = _validate_transformers_model_dict(transformers_model)

# This allows us to do:

transformers_model.model

# Instead of:
transformers_model[_MODEL_KEY]  # this looks unsafe
```

Reply:
great idea! I didn't use the typing.NamedTuple subclassing due to the limitations of defining an indeterminate field naming from a key-optional dict input and instead used the function-based input with dynamic field naming from collections.namedtuple. The weird behavior of `__new__()` instead of `__init__()` in NamedTuple would have meant subclassing a subclass in order to get a classmethod available that could parse and populate the NamedTuple instance (I tried it and the code was getting pretty nasty looking)